src/otx/algo/samplers/balanced_sampler.py

```diff
@@ -0,0 +1,174 @@
+        n_repeats: int = 1,
+        generator: torch.Generator | None = None,
+    ):
+        self.dataset = dataset
+        self.samples_per_gpu = samples_per_gpu
+        self.num_replicas = num_replicas
+        self.rank = rank
+        self.drop_last = drop_last
+        self.generator = generator
+        self.repeat = n_repeats
+
+        super().__init__(dataset)
+
+        # img_indices: dict[label: list[idx]]
+        ann_stats = get_idx_list_per_classes(dataset.dm_subset)
+        self.img_indices = {k: v for k, v in ann_stats.items() if len(v) > 0}
+        self.num_cls = len(self.img_indices.keys())
+        self.data_length = len(self.dataset)
+        self.num_trials = int(self.data_length / self.num_cls)
+
+        if efficient_mode:
+            # Reduce the # of sampling (sampling data for a single epoch)
+            num_tail = min(len(cls_indices) for cls_indices in self.img_indices.values())
+            if num_tail > 1:
+                base = 1 - (1 / num_tail)
+                num_reduced_trials = int(math.log(0.001, base))
+                self.num_trials = min(num_reduced_trials, self.num_trials)
+
+        self.num_samples = self._calculate_num_samples()
+
+    def _calculate_num_samples(self) -> int:
+        num_samples = self.num_trials * self.num_cls * self.repeat
+
+        if self.num_replicas > 1:
+            # If the dataset length is evenly divisible by # of replicas, then there
+            # is no need to drop any data, since the dataset will be split equally.
+            if self.drop_last and num_samples % self.num_replicas != 0:
+                # Split to nearest available length that is evenly divisible.
+                # This is to ensure each rank receives the same amount of data when
+                # using this Sampler.
+                num_samples = math.ceil(
+                    # `type:ignore` is required because Dataset cannot provide a default __len__
+                    # see NOTE in pytorch/torch/utils/data/sampler.py
+                    (num_samples - self.num_replicas) / self.num_replicas,
+                )
+            else:
+                num_samples = math.ceil(num_samples / self.num_replicas)
+            self.total_size = num_samples * self.num_replicas
+
+        return num_samples
+
+    def __iter__(self):
+        """Iter."""
+        if self.generator is None:
+            seed = int(torch.empty((), dtype=torch.int64).random_().item())
+            generator = torch.Generator()
+            generator.manual_seed(seed)
+        else:
+            generator = self.generator
+
+        indices = []
+        for _ in range(self.repeat):
+            for _ in range(self.num_trials):
+                indice = torch.cat(
+                    [
+                        torch.tensor(self.img_indices[cls_indices], dtype=torch.int64)[
+                            torch.multinomial(torch.ones(len(self.img_indices[cls_indices])), 1, generator=generator)
```

Comment:
As the same reason as https://github.com/openvinotoolkit/training_extensions/pull/2993/files#r1505322508, `torch.multinomial(torch.ones(len(self.img_indices[cls_indices]))` is inefficient. Use `torch.randint` rather than creating `torch.ones(len(self.img_indices[cls_indices])` and using `torch.multinomial()`. In addition, it is inefficient just sample one index each time. Find another way to sample a batch of indices to make it faster such as 
```python
        if self.replacement:
            for _ in range(self.num_samples // 32):
                yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=generator).tolist()
            yield from torch.randint(high=n, size=(self.num_samples % 32,), dtype=torch.int64, generator=generator).tolist()
```

Reply:
Thanks for the nice comment, I replaced `torch.multinomial(torch.ones(len(self.img_indices[cls_indices]))` with `torch.randint(0, len(self.img_indices[cls_indices]), (1,), generator=self.generator)` as you suggested.
However, I've seen suggestions on how to sample a batch of indices,
I haven't found any code yet that provides batch in a more efficient way, I'll look into it some more.

Reply:
Improve it further to sample a bulk of random indices in the next PR. Refer to
```python
        if self.replacement:
            for _ in range(self.num_samples // 32):
                yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=generator).tolist()
            yield from torch.randint(high=n, size=(self.num_samples % 32,), dtype=torch.int64, generator=generator).tolist()
```

Reply:
Yes I was thinking about this now, I will change this next time, thank you.